kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/model/BytesOutputRow.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 Lenses.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.lenses.streamreactor.connect.aws.s3.model
+
+import java.io.InputStream
+
+import io.lenses.streamreactor.connect.aws.s3.config.BytesWriteMode
+import io.lenses.streamreactor.connect.aws.s3.config.BytesWriteMode._
+
+import scala.collection.mutable.ListBuffer
+
+object BytesOutputRow {
+
+  def apply(storedByteArray: Array[Byte], bytesWriteMode: BytesWriteMode): BytesOutputRow = {
+
+    if (bytesWriteMode == KeyOnly) {
+      new BytesOutputRow(None, None, storedByteArray, Array.empty)
+
+    } else if (bytesWriteMode == ValueOnly) {
+      new BytesOutputRow(None, None, Array.empty, storedByteArray)
+
+    } else {
+      throw new IllegalArgumentException(s"Invalid apply function for bytesWriteMode key/value only $bytesWriteMode")
+    }
+  }
+
+  def apply(inputStream: InputStream, bytesWriteMode: BytesWriteMode): BytesOutputRow = {
```

Comment:
It's cleaner to create a reader from byteswritemode which takes the InputStream and avoid the if else. The logic of how it is read is then in the implementation. Easier to test, easier to extend.

Reply:
I have done what you suggest, it has cut down the logic a lot, very nice suggestion.